src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -293,23 +297,35 @@ private void OnWriteCompleted(WriteContext writeContext)
             {
                 PoolWriteContext(writeContext);
 
-                // _numBytesPreCompleted can temporarily go negative in the event there are
-                // completed writes that we haven't triggered callbacks for yet.
-                _numBytesPreCompleted -= bytesWritten;
-
-                // bytesLeftToBuffer can be greater than _maxBytesPreCompleted
-                // This allows large writes to complete once they've actually finished.
-                var bytesLeftToBuffer = _maxBytesPreCompleted - _numBytesPreCompleted;
-                while (_tasksPending.Count > 0 &&
-                       (int)(_tasksPending.Peek().Task.AsyncState) <= bytesLeftToBuffer)
+                if (_lastWriteError == null)
                 {
-                    var tcs = _tasksPending.Dequeue();
-                    var bytesToWrite = (int)tcs.Task.AsyncState;
+                    // _numBytesPreCompleted can temporarily go negative in the event there are
+                    // completed writes that we haven't triggered callbacks for yet.
+                    _numBytesPreCompleted -= bytesWritten;
+
+                    // bytesLeftToBuffer can be greater than _maxBytesPreCompleted
+                    // This allows large writes to complete once they've actually finished.
+                    var bytesLeftToBuffer = _maxBytesPreCompleted - _numBytesPreCompleted;
+                    while (_tasksPending.Count > 0 &&
+                           (int)(_tasksPending.Peek().Task.AsyncState) <= bytesLeftToBuffer)
+                    {
+                        var tcs = _tasksPending.Dequeue();
+                        var bytesToWrite = (int)tcs.Task.AsyncState;
 
-                    _numBytesPreCompleted += bytesToWrite;
-                    bytesLeftToBuffer -= bytesToWrite;
+                        _numBytesPreCompleted += bytesToWrite;
+                        bytesLeftToBuffer -= bytesToWrite;
 
-                    _tasksCompleted.Enqueue(tcs);
+                        _tasksCompleted.Enqueue(tcs);
+                    }
+                }
+                else
+                {
+                    // Error all waiting tasks
+                    while (_tasksPending.Count > 0)
```

Comment:
Thanks for doing this. I agree with this approach. This is type of change is what @rynowak and I were considering when he filed the issues at the end of last week.

I think we also want to fail all pending tasks if `OnWriteCompleted` is called after a `SocketDisconnect` just to be safe.

Another consideration is how this is going to interact with the per-request [IgnoreWriteExceptions](https://github.com/aspnet/KestrelHttpServer/issues/566#issuecomment-170194556) property we are likely to add. We still haven't decided whether or not it will be opt in or out, but I'm personally leaning toward opt-in.

This PR also conflicts a little with https://github.com/aspnet/KestrelHttpServer/pull/562 so one of us is going to have to rebase depending on the order these get in (and we know who controls that :stuck_out_tongue:). Seriously though, I'm going to try to merge #562 today and the rebase shouldn't be too difficult.